### PR TITLE
Ref #25491 - Add public key option in create compute resource CLI

### PR DESCRIPTION
**Pull request by shiramax**

### CODE REVIEW

**Comment by mbacovsky** [comment]
@shiramax, thanks for the patch It looks good, could you please add a test for it?

**Comment by shiramax** [comment]
thanks, @mbacovsky , please review again.

**Comment by mbacovsky** [comment]
Please rebase the PR to work with latest core.

File: lib/hammer_cli_foreman/compute_resource.rb [hunk 1]
```diff
@@ -70,6 +70,11 @@ def print_data(data)
 
     class CreateCommand < HammerCLIForeman::CreateCommand
 
+
+      option "--public_key_path", "PUBLIC_KEY_PATH", _("Path to a file that contains oVirt public key (For oVirt only)"), :attribute_name => :option_public_key_path,
```
[the remark this answers is not in the document]
Comment (shiramax) [reply lib/hammer_cli_foreman/compute_resource.rb]: It seems that there are multiple parameters that are relevant to only one compute resource, I guess the reason is that we in the future we will able to add a compute resource with the same parameters.

**Comment by shiramax** [comment]
@mbacovsky thanks :)
please review again.

File: lib/hammer_cli_foreman/compute_resource.rb (hunk 1 above)
Comment (mbacovsky) [inline lib/hammer_cli_foreman/compute_resource.rb]: If you use `:attribute_name => :option_public_key` you can remove `request_params` definition completely.

File: lib/hammer_cli_foreman/compute_resource.rb [hunk 2]
```diff
@@ -82,6 +87,13 @@ class CreateCommand < HammerCLIForeman::CreateCommand
...
       failure_message _("Could not create the compute resource")
 
+      def request_params
+        params = super
+        if options["option_public_key_path"]
+          params['compute_resource']['public_key'] = options["option_public_key_path"]
+        end
+        params
+      end
       build_options
```
Comment (mbacovsky) [inline lib/hammer_cli_foreman/compute_resource.rb]: In the command help I can see:
```
 --public-key PUBLIC_KEY                     For oVirt only
 --public_key_path PUBLIC_KEY_PATH           Path to a file that contains oVirt public key (For oVirt only)
```
This is probably unwanted. Use `build_options :without => :public_key`

File: lib/hammer_cli_foreman/compute_resource.rb (hunk 1 above)
Comment (mbacovsky) [inline lib/hammer_cli_foreman/compute_resource.rb]: In `--public_key_path` use dashes as in `--public-key-path`

Comment (shiramax) [reply lib/hammer_cli_foreman/compute_resource.rb]: I tried it, but it didn't work - the public_key was not in params['compute_resource']

**Comment by ofedoren** [comment]
Just few additional notes:

-  `:required => false` is not necessary and can be omitted.
-  if you specify `attribute_name => :option_public_key` then it puts the key value under `['compute_resource']` in params and `request_params` section is redundant. At least I see it in debug mode and also in the request.
-  also we need to add `--public-key-path` option to `UpdateCommand` as well to be consistent with `CreateCommand` (as mbacovsky said). And a test for it as well.
-  currently, if a path was specified with `--public-key-path` option, the specified path will be sent to the server as param instead of value in that file. While I was debugging, I noticed that when you specify `attribute_name => :option_public_key` then the `File` (or any) normalizer doesn't do its job (should we open an issue about that, @mbacovsky? It's weird, because the tests pass). I guess that's why the path is sent instead of file content.

Also I've noticed that the tests fail if there is no `request_params` section specified, which is weird also, because the actual request contains `['compute_resource']['public_key']` param. Not sure now where is the problem, in tests or core.

**Comment by mbacovsky** [comment]
@shiramax, @ofedoren I did some debugging of what is going on with  `attribute_name => :option_public_key`. My findings are:
- when clamp defines an option it creates writer method for it that is used later by the parser. The writer method is called `#{attribute_name}=` and in the body the normalizer as called and the value is assigned to the attribute in the command.
- during parsing clamp finds the matching option definition and asks for the writer name and calls that method with the parsed value
- in our case --public-key-path will create writer `option_public_key=` which is later re-defined by writer for --public-key. That is the reason why the File normalizer is not triggered.
- it seems there is not a nice fix for this general behavior and we should better avoid using :attribute_name same as other existing option of different type

One solution here could be to define a new normalizer FileOrValue that will check if the value is existing file and loads its content  or use the value provided. In that case we would have to :exclude public_key and define our own with the proper normalizer and we could get rid of requiest_params modifications.
Other solution is to keep the two options as suggested in the PR but we have to fix the request params 
```ruby
params['compute_resource']['public_key'] = option_public_key_path if option_public_key_path
```
otherwise the public key is overriden with nil. 

I'm okay both ways.

What would you prefer @shiramax?

**Comment by shiramax** [comment]
Thanks @mbacovsky and @ofedoren.
I fixed the request params, 
thanks!

File: lib/hammer_cli_foreman/compute_resource.rb [hunk 3]
```diff
@@ -71,12 +71,21 @@ def print_data(data)
 
 
     class CreateCommand < HammerCLIForeman::CreateCommand
+      option "--public-key-path", "PUBLIC_KEY_PATH", _("Path to a file that contains oVirt public key (For oVirt only)"), :attribute_name => :option_public_key_path,
+             :required => false, :format => HammerCLI::Options::Normalizers::File.new
```
Comment (mbacovsky) [inline lib/hammer_cli_foreman/compute_resource.rb]: I wonder why the `:required => false` and `:attribute_name => :option_public_key_path` are kept in the code. These are the default values and should have no impact on the code. Just making it more difficult to read. Is it intentional?

**Comment by mbacovsky** [comment]
Excelent! Thanks @shiramax!

**Pull request merged**